Add XenVM dom0 virtual shield and prepare/cleanup r8a7795 configs/code for working with the new shield

xakep-amatop opened the pull request.

xakep-amatop committed:
dts: arm64: use 2-words cells for reg and size inside r8a7795

Use 2 words for 'address-cells' and 'size-cells', because we have 64-bit
architecture and we want to make it easy applicable with XenVM initial
domain shield overlay;

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
soc: arm64: xenvm: move XEN_INITIAL_DOMAIN config to Xen driver

Untie XEN_INITIAL_DOMAIN from XenVM SoC configuration,
it is necessary for running xen dom-0 on different boards.

Move XEN_INITIAL_DOMAIN from XenVM SoC configuration to
Xen driver.

Set XEN_INITIAL_DOMAIN by default to n in order to prevent issues
with existing setups, which may use XenVM as DomU.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
drivers: uart_rcar: do MMIO mapping inside driver

Add MMIO mapping for UART Renesas driver in order to avoid
mappings inside mmu_regions.c file(s) and remove MMU region
for Renesas SCIF driver, because driver uses mapping inside
init driver function.

There are a lot of changes inside SCIF driver inside this
commit, because reg addr and size may be stored in RAM or
ROM and appropriately in different driver structures data
or cfg, and, because the name of field reg base is changed.

Note: it is common approach according to Zephyr documentation.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
drivers: pfc_rcar: do MMIO mapping inside driver

Add MMIO mapping for PFC Renesas driver in order to avoid
mappings inside mmu_regions.c file. Remove MMU region for
Renesas PFC driver, because driver uses a run-time mapping
inside init driver function.

Delete ROUND_UP_SIZE_MMU_PAGE_FOR_NODE macro, because it
isn't used anymore.

Add a new system init function pfc_rcar_driver_init to PFC
Renesas driver for invoking a memory mapping macro.
Note: PFC Renesas driver doesn't use Zephyr Device Model.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
dts: bindings: xen: add xen,xen.yaml file

Add yaml file for 'xen,xen', because without it an appropriate
'CONFIG_DT_HAS_XEN_XEN_ENABLED' isn't generated.

Include a new generic xen,xen.yaml file to xen,xen-4.15.yaml,
because the first has properties for both.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop commented on boards/shields/xen_dom0_rcar/doc/index.rst:
(note) I'll add documentation in a separate commit

xakep-amatop commented:
there are some errors in GitHub tests
pls, postpone review of this pull until I fix them

xakep-amatop commented:
> there are some errors in GitHub tests pls, postpone review of this pull until I fix them

ready for review

lorc commented on drivers/xen/Kconfig:
```diff
@@ -5,6 +5,12 @@ if SOC_XENVM
 
+config XEN_INITIAL_DOMAIN
+	bool "Zephyr as Xen Domain 0"
+	default y
```
I'd prefer to have it `n` by default. This is to prevent issues with existing setup, which my use `xenvm` as DomU

xakep-amatop replied:
good point! I'll change it to `n`

lorc commented on soc/arm64/renesas_rcar/gen3/mmu_regions.c:
```diff
@@ -21,13 +21,6 @@ static const struct arm_mmu_region mmu_regions[] = {
 			      DT_REG_SIZE_BY_IDX(DT_INST(0, arm_gic), 1),
 			      MT_DEVICE_nGnRnE | MT_RW | MT_NS),
 
-#if DT_NODE_HAS_STATUS(DT_INST(0, renesas_rcar_scif), okay)
```
Shouldn't you fold this change into "drivers: uart_rcar: do runtime MMIO mapping"?

I am asking, because your previous patch can break Zephyr is applied without this change here.

xakep-amatop replied:
ok, I'll squash commits together

lorc commented on soc/arm64/renesas_rcar/gen3/mmu_regions.c:
```diff
@@ -26,11 +23,12 @@ static const struct arm_mmu_region mmu_regions[] = {
 			      DT_REG_SIZE(DT_INST(0, renesas_r8a7795_cpg_mssr)),
 			      MT_DEVICE_nGnRnE | MT_RW | MT_NS),
 
-#if DT_NODE_HAS_STATUS(DT_INST(0, renesas_rcar_pfc), okay)
```
I believe that this change should belong to previous commit.

lorc replied:
As I said, this change should be squashed to the previous commit

xakep-amatop replied:
done

lorc commented on arch/arm64/core/CMakeLists.txt:
```diff
@@ -53,4 +53,6 @@ if(CMAKE_C_COMPILER_ID STREQUAL "GNU")
 endif()
 
-add_subdirectory_ifdef(CONFIG_SOC_XENVM xen)
+if (CONFIG_SOC_XENVM OR CONFIG_DT_HAS_XEN_XEN_4_15_ENABLED)
```
Ugh. Why do you hardcoding Xen version there?  

For this exact reason, there are two separate compatible entries: `compatible = "xen,xen-4.15", "xen,xen";` please use the second one if there is no hard dependency on this particular Xen version. 

Also, please take into account that 4.15 is not an actual Xen version running on a platform.

xakep-amatop replied:
because CONFIG_DT_HAS_XEN_XEN_ENABLED wasn't defined during build and GitHub automation notified about that
if you want I can add extra check here too

firscity replied:
Probably Zephyr DT parser does not add all compatibles to these defines. I think it will be okay to remove 4.15 from DT compatibles at all. All Zephyr drivers uses `xen,xen`. I don't think that we should update exact version every time.

lorc commented on boards/shields/xenvm_dom0/Kconfig.defconfig:
```diff
@@ -0,0 +1,56 @@
+
+if SHIELD_XENVM_DOM0
+
+config NUM_IRQS
```
Is there reason why you can't use value from original SoC/Board?

xakep-amatop replied:
different number of IRQ for h3ulcb and XENVM boards, e.g. 240 vs 500

lorc commented on boards/shields/xenvm_dom0/Kconfig.defconfig:
```diff
@@ -0,0 +1,56 @@
+config UART_XEN_HVC_CONSOLEIO
+	default y
+
+config TEST
```
Do you really need to enable this for all builds?

xakep-amatop replied:
sorry, I didn't investigate this and below questions, I have just copy-pasted a part of XENVM board config

lorc commented on boards/shields/xenvm_dom0/Kconfig.defconfig:
```diff
@@ -0,0 +1,56 @@
+config SHELL_BACKEND_SERIAL
+	default y
+
+config DEBUG_INFO
```
Is there reason why you can't use value from original SoC/Board?

lorc commented on boards/shields/xenvm_dom0/Kconfig.defconfig:
```diff
@@ -0,0 +1,56 @@
+config DEBUG_INFO
+	default y
+
+config OVERRIDE_FRAME_POINTER_DEFAULT
```
Is there reason why you can't use value from original SoC/Board?

lorc commented on boards/shields/xenvm_dom0/Kconfig.defconfig:
```diff
@@ -0,0 +1,56 @@
+config OVERRIDE_FRAME_POINTER_DEFAULT
+	default y
+
+config OMIT_FRAME_POINTER
```
Is there reason why you can't use value from original SoC/Board?

lorc commented on boards/shields/xenvm_dom0/Kconfig.defconfig:
```diff
@@ -0,0 +1,56 @@
+config SHELL_STACK_SIZE
+	default 32768
+
+endif # SHIELD_XENVM_DOM0
```
I believe that you have valid reason to define all the stack sizes there

firscity replied:
I think that we need to add stack changes with `TODO:` as they were added to xenvm_defconfig. No one investigated exact value, that we need, it was just a WA for Gen3 boards.

lorc commented on boards/shields/xenvm_dom0/doc/index.rst:
```diff
@@ -0,0 +1,29 @@
+********
+
+This is a virtual shield that allows you to use any board config as a Xen
+initial domain.
```
Strange wording. You can use "board config" as "initial domain". "  How about "This virtual shield allows user to build Zephyr as a Xen initial domain (Dom0). This feature is implemented as shield to allow support for any compatible platform".

lorc commented on boards/shields/xenvm_dom0/Kconfig.shield:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SHIELD_XENVM_DOM0
+	def_bool $(shields_list_contains,xenvm_dom0)
```
Could you please add `depends on ARMV8_A`? I don't believe that this would work on CorteR or MIPS :)

xakep-amatop replied:
done

lorc commented on boards/shields/xenvm_dom0/xenvm_dom0.overlay:
```diff
@@ -0,0 +1,39 @@
+		method = "hvc";
+	};
+
+	ram: memory@60000000 {
```
RAM base and RAM size should be platform-dependent.

lorc reviewed (commented):
This looks quite good. Thank you. But I want to ask you for two things:

- Be sure that commits are atomic. Among other things this means that I can do `git checkout` at any commit in the PR and nothing will break.
- Provide meaningful commit descriptions.  `boards: arm64: h3ulcb: add ram node to chosen` is the largest offender. I saw the changes, I saw the commit description. Looks like you are doing right thing, but I have no idea why you are doing this and how to check if this is really correct way. Please provide elaborated commit descriptions for every commit.

firscity commented on soc/arm64/xenvm/Kconfig.soc:
```diff
@@ -8,9 +8,3 @@ config SOC_XENVM
 	select GIC_V2 if !XENVM_USE_GIC_V3
 	select CPU_CORTEX_A72
-
-config XEN_INITIAL_DOMAIN
```
This commit will broke build, please add only atomic changes. All platforms (xenvm, xenvm-qemu should be buildable)

xakep-amatop replied:
done

firscity commented on boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57_defconfig:
```diff
@@ -25,9 +19,7 @@ CONFIG_MMC_RCA=1
 
 # Enable console
 CONFIG_CONSOLE=y
-CONFIG_UART_RCAR=y
```
Don't think that these changes should belong to commit with such description.

xakep-amatop replied:
I have extended commit message, thanks

lorc replied:
Commit name still says "dts: arm64: changes for r8a7795.dtsi".

Also, please take note that this is too broad for the commit name. Any patch can be described as "changes to some_file.ext"

xakep-amatop replied:
as I have written I changed commit message and @firscity has said that changes isn't related to commit description
nobody has said nothing about commit name

lorc replied:
Commit name is a part of the commit description as a whole. Just by looking at your commit (for example when doing `git log --oneline`) tells me that you are touching only r8a7795.dtsi. Which is not true.

xakep-amatop committed:
soc: arm64: rcar: mmu: add mapping of Hypervisor mem region

Add a MMU mapping of Xen Hypervisor memory. It is needed for
using Rcar Gen3 SoC for Xen Initial domain.

Note: mapping will occur only in case if node present in dts
      ant it has 'okay' status.

Todo: add mapping to a init driver function or use SYS_INIT
      for it in case if driver init function doesn't exist.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop commented:
fix everything except this comment https://github.com/xen-troops/zephyr/pull/13/files#r1137772562
working on solution...

dsemenets commented:
Looks like you forget https://github.com/xen-troops/zephyr/pull/13/files#r1137782692

xakep-amatop commented:
> Looks like you forget https://github.com/xen-troops/zephyr/pull/13/files#r1137782692

@dsemenets you are right, thanks
looks like, I have deleted it accidentally during cleanup

lorc reviewed (commented):
There are lots of semi-related commits in this PR, which makes it hard to review.

Some of them need to be squashed together. Other - should be split up.  `boards: shields: add XEN Dom-0 shield` for example adds new DTS binding. Why are you doing this in this particular commit?

I propose to split this PR into couple of separate ones. For example, changes to PFC, UART and CPG drivers are not Xen-related and almost already merged. Just move them to a separate PR. Xen shield can go into another.

Funny thing is that PR is named   `xt rcar mmc driver integration xenvm` but there are no MMC-related changes at all :)

xakep-amatop commented:
@lorc 

> Some of them need to be squashed together.

agree, there is one commit remaining that can be squashed

> Other - should be split up. `boards: shields: add XEN Dom-0 shield` for example adds new DTS binding. Why are you doing this in this particular commit?

because all stuff inside the last commit related to adding of a new shield, I see no reason for splitting it up 

> I propose to split this PR into couple of separate ones. For example, changes to PFC, UART and CPG drivers are not Xen-related and almost already merged. Just move them to a separate PR. Xen shield can go into another.

I think, it is better to leave the pull as is, because it will require more unnecessary time wasting from me and you, you need time for carefully review of everything (again) and I need time for creating another pull, moreover changes distinguish from one another by commits and  all commits will have the same appearance as after merge from different pulls as after merge only the one pull

> Funny thing is that PR is named xt rcar mmc driver integration xenvm but there are no MMC-related changes at all :)

it is historical name, content of pull-request was changed a lot during development process

lorc commented:
> because all stuff inside the last commit related to adding of a new shield, I see no reason for splitting it up

You can go further and squash all commits into one :) 

Actually no. You need to move unrelated changes to a separate commits. For example, DTS schema may be picked up by Zephyr maintainer separately from Xen shield.

> I think, it is better to leave the pull as is, because it will require more unnecessary time wasting from me and you, you need time for carefully review of everything (again) and I need time for creating another pull, moreover changes distinguish from one another by commits and all commits will have the same appearance as after merge from different pulls as after merge only the one pull

Well, I am not forcing you. This was just my idea to speed up the review process.

> it is historical name, content of pull-request was changed a lot during development process

FYI, you can rename the PR at any time.

xakep-amatop commented:
> Actually no. You need to move unrelated changes to a separate commits. For example, DTS schema may be picked up by Zephyr maintainer separately from Xen shield.

any part of commit may be picked by anyone...

> Well, I am not forcing you. This was just my idea to speed up the review process.
 
as far as I know you and Dmytro have already reviewed all my changes, how it can speed up review process... sorry, but I have strong feeling that you are laughing at me

lorc reviewed (commented):
> any part of commit may be picked by anyone...

This is technically true. But it is customary to cherry-pick commits as a whole. This helps preserving code authorship. Remember, that commits should be atomic. One commit should do one specific thing. This eases up and speeds up the review process and leads to a much cleaner and well understood history.  Also, you easily can revert some of the changes, if they cause issues. 

> as far as I know you and Dmytro have already reviewed all my changes, how it can speed up review process... sorry, but I have strong feeling that you are laughing at me

I am sorry that you have this feeling. I didn't meant harm. But, from my point of view, there are lot to do in some of the commits. As you can see, there are still ongoing conversations regarding on some of the changes. 

Also, I want to ask you to review again the following two sections:

https://www.kernel.org/doc/html/latest/process/submitting-patches.html#split-changes

https://www.kernel.org/doc/html/latest/process/submitting-patches.html#describe-your-changes

Those are the general guidelines that we (and all the open-source community) trying to follow.

Pay attention to the next statements: 

> "The point to remember is that each patch should make an easily understood change that can be verified by reviewers. Each patch should be justifiable on its own merits." 

> "Solve only one problem per patch. If your description starts to get long, that’s a sign that you probably need to split up your patch." 

> "Describe your changes in imperative mood, e.g. “make xyzzy do frotz” instead of “[This patch] makes xyzzy do frotz” or “[I] changed xyzzy to do frotz”, as if you are giving orders to the codebase to change its behaviour."

xakep-amatop committed:
shields: xen_dom0: add overlay for h3ulcb board

Delete ram and redefine memory node for h3ulcb board, because
reg address and size can be different for bar-metal and XenVM
dom0 builds;

Delete scif node for h3ulcb board, because dom0 uses Xen
Hypervisor takes serial and uses own node for this.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
dts: arm64: move gic node outside soc node inside r8a7795

Move gic node outside soc node. I haven't found any binding rules
which can say that it is should be like this, but we use the same
approach for a lot of boards inside arm64 dir, so I suppose there
isn't mistake here. Moreover, I want to sync this dtsi as much as
possible with XenVM dts because in Xen Initial domain shield will
use some of nodes and their properties from XenVM board dts.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
boards: arm64: add memory node to h3ulcb

Add memory node to the h3ulcb board. After that we can delete
all of hardcoded memory configs from the board default configuration,
they can be correctly loaded if we add 'zephyr,sram' property to
chosen node.

Also, there was discrepancy between YAML configuration and DTS/cfg
values, so I've synced all values between one another and increased
their size to 512M in order to avoid further expansion in the near
future.

Note: the board has 8GB of RAM.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
boards: arm64: cleanup defconfig for h3ulcb

ARM_ARCH_TIMER and CLOCK_CONTROL_RCAR_CPG_MSSR are automatically chosen
by SoC default configuration of Renesas Gen3.

UART_RCAR is automatically selected if there is a SCIF node inside dts.

We don't need FLASH configs, because they should be loaded automatically
from appropriate chosen node.

Logging system and MMC driver related configs should be disabled by
default (board def configs isn't a right place for doing it) and only
app config should enable it in case of need.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
dts: arm64: add PSCI node to r8a7795

Add PSCI node to r8a7795 in order to use a standard interface
for power management.
Note: currently driver supports only 3 actions:
      * CPU on;
      * CPU off;
      * System off.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
drivers: add subdirectory of xen when we have Xen Hypervisor node

Add subdirectory of 'xen' when we have Xen Hypervisor node enabled
in dts. It is necessary action if we want to use driver for Xen
Initial domain shield, that means that the driver isn't tied only
to XenVM board configuration.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
arch: arm64: add subdirectory of xen when we have Xen Hypervisor node

Add subdirectory of 'xen' when we have Xen Hypervisor node enabled
in dts. It is necessary action if we want to use arch arm64 core Xen
related code for Xen Initial domain shield, that means that the core
of Xen isn't tied only to XenVM SoC configuration.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
boards: shields: add virtual Xen dom0 shield support

Add virtual Xen dom0 shield support and the dependence of
configs XEN_EARLY_CONSOLEIO and UART_XEN_HVC_CONSOLEIO on
configuration of shield 'xenvm_dom0'.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
tests: add h3ulcb configs for running some tests on MMC device

Add MMC related configs to disk/MMC tests for supporting h3ulcb board:
* tests/drivers/disk/disk_performance;
* tests/subsys/sd/mmc.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

lorc commented on boards/arm64/rcar_h3ulcb_ca57/rcar_h3ulcb_ca57.yaml:
```diff
@@ -5,7 +5,7 @@ arch: arm64
   - zephyr
   - cross-compile
-ram: 256
+ram: 512
```
commit description mentions only 16MB... I do understand why you are making this change:  our board configuration was really inconsistent, but it is worth mentioning in the commit message.

xakep-amatop replied:
done

lorc reviewed (approved):
Thank you. It is much easier to review and it is better structured overall.

I have left last comment. Please address it and then add my tag to all the commits:

`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

firscity reviewed (approved):
Good job, thank you, please add
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`
